### ISSUE TITLE
Half of the tests fail after failing to open /sys/fs/ghost/ctl

### ISSUE THREAD

**Issue by Swire42** [google/ghost-userspace#21]
Hi,

When running `# bazel test ...` on my computers, half of the tests fail, for instance `api_test` or `agent_test`.

I made a script to help reproduce my exact kernel install process, so you can try that if you are not able to reproduce:
- Install a fresh Ubuntu 20.04 from a usb stick
- run the bash command found at https://github.com/Swire42/ghost-installer/blob/master/README.md

Here is what `test.log` looks like:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //:agent_test
-----------------------------------------------------------------------------
Running main() from gmock_main.cc
[==========] Running 13 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 13 tests from AgentTest
[ RUN      ] AgentTest.DestructorCanFree
lib/enclave.cc:420(12) CHECK FAILED: ctl_fd_ >= 0 [-1 > 0]
errno: 30 [Read-only file system]
PID 139761 Backtrace:
[0] 0x7f58edc5a59f : ghost::Exit()
[1] 0x7f58edec3d7a : ghost::LocalEnclave::CreateAndAttachToEnclave()
[2] 0x7f58edec484d : ghost::LocalEnclave::LocalEnclave()
[3] 0x55d62a146d43 : ghost::FullAgent<>::FullAgent()
[4] 0x55d62a1319c2 : ghost::(anonymous namespace)::FullSimpleAgent<>::FullSimpleAgent()
[5] 0x55d62a12e18e : std::make_unique<>()
[6] 0x55d62a12b230 : ghost::AgentProcess<>::AgentProcess()
[7] 0x55d62a1240d9 : ghost::(anonymous namespace)::AgentTest_DestructorCanFree_Test::TestBody()
[8] 0x7f58ed8d72ca : testing::internal::HandleSehExceptionsInMethodIfSupported<>()
[9] 0x7f58ed8d1e09 : testing::internal::HandleExceptionsInMethodIfSupported<>()
[10] 0x7f58ed8b90f6 : testing::Test::Run()
[11] 0x7f58ed8b9acd : testing::TestInfo::Run()
[12] 0x7f58ed8ba389 : testing::TestSuite::Run()
[13] 0x7f58ed8c6ab5 : testing::internal::UnitTestImpl::RunAllTests()
[14] 0x7f58ed8d81ea : testing::internal::HandleSehExceptionsInMethodIfSupported<>()
[15] 0x7f58ed8d2fb1 : testing::internal::HandleExceptionsInMethodIfSupported<>()
[16] 0x7f58ed8c52a7 : testing::UnitTest::Run()
[17] 0x7f58ed921af8 : RUN_ALL_TESTS()
[18] 0x7f58ed921a7a : main
[19] 0x7f58ed235083 : __libc_start_main
```

The error seems to occur in `LocalEnclave::MakeNextEnclave()` at `lib/enclave.cc:177`, when opening `/sys/fs/ghost/ctl`.

I have no idea as to why the error is `Read-only file system`, given that the output of `ls -l /sys/fs/ghost/ctl` is:
```
-rw-rw---- 1 root root 0 juin  30 15:56 /sys/fs/ghost/ctl
```
(Note: I am running the tests with root privileges: `sudo su` then `bazel test ...`)

Seeing this, the problem might come from the kernel.

I hope you will be able to figure out what is wrong here.
These errors are not a fundamental problem to me as agents are able to run and schedule tasks without problems, but it would be nice to pass tests nonetheless.

**Comment by jackhumphries** [comment]
Are you still experiencing this issue? Please let me know.

**Comment by Swire42** [comment]
This issue still happens with the most recent ghost kernel and userspace. 

Steps to reproduce :
1. Install a fresh new Ubuntu 20.04 
2. Run https://github.com/Swire42/ghost-installer (replicates my install process)
3. Clone https://github.com/google/ghost-userspace
4. Run "bazel test ..." within it, as root 

(optionnal : if needed, fix compile time errors such as c++20 not being available in Ubuntu 20.04)

**Comment by jackhumphries** [comment]
Is `/sys` mounted as a read-only file system?

**Comment by Swire42** [comment]
I don't think so, for instance I am able to destroy enclave 1 by writing "destroy" in /sys/fs/ghost/enclave_1/ctl

**Comment by jackhumphries** [comment]
This is tricky to debug because ghOSt itself does not return that error. Rather, it's coming from somewhere within the standard Linux kernel code.

As a quick check, what happens if you run the test binary directly (e.g., `bazel-bin/agent_test`) instead of with `blaze test`?

**Comment by Swire42** [comment]
Note : there seems to be no problem with /sys/fs/ghost/ctl when running things manually:
```
# ls /sys/fs/ghost/
ctl  version
# echo create 1 69 >> /sys/fs/ghost/ctl 
# ls /sys/fs/ghost/
ctl  enclave_1  version
# echo destroy >> /sys/fs/ghost/enclave_1/ctl
# ls /sys/fs/ghost/
ctl  version
```

**Comment by jackhumphries** [comment]
Alright, let me look into this. It is likely that `bazel test` is launching the code in some sort of container that prohibits write access to `/sys`.

**Comment by Swire42** [comment]
Running the test binary directly as you suggests results in a success

**Comment by Swire42** [comment]
Is it possible that somehow Bazel does not run the tests as root despite Bazel itself being run as root ?

**Comment by jackhumphries** [comment]
`bazel test` [sandboxes](https://bazel.build/docs/sandboxing) tests which is why the tests are unable to get write access to `/sys`. The tests need write access to `/sys/fs/ghost` so that they can communicate with the ghOSt kernel code.

To turn off sandboxing, you can either run the test binaries directly or you can invoke `bazel test` with the `--spawn_strategy local` argument.

Please let me know if this works. If so, I will update the README to explain this. Thanks!

**Comment by Swire42** [comment]
`bazel test --spawn_strategy local agent_test` works fine, this issue is therefore solved.

However, there seem to be a lot of tests that result in timeouts on my machine.
I am not sure if it is due to the limited capacities of my machine or to broken tests.
Examples of tests that timeout: api_test, biff_test, database_test, ...

Additionally enclave_test encounters problems:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //:enclave_test
-----------------------------------------------------------------------------
Running main() from gmock_main.cc
[==========] Running 10 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 10 tests from EnclaveTest
[ RUN      ] EnclaveTest.TxnRegionFree
[       OK ] EnclaveTest.TxnRegionFree (47 ms)
[ RUN      ] EnclaveTest.MakeAnyEnclave
[       OK ] EnclaveTest.MakeAnyEnclave (1 ms)
[ RUN      ] EnclaveTest.SetCpus
[       OK ] EnclaveTest.SetCpus (275 ms)
[ RUN      ] EnclaveTest.BadCpus
[       OK ] EnclaveTest.BadCpus (0 ms)
[ RUN      ] EnclaveTest.CleanupLeakedCpus
[       OK ] EnclaveTest.CleanupLeakedCpus (39 ms)
[ RUN      ] EnclaveTest.MmapTxn
[       OK ] EnclaveTest.MmapTxn (32 ms)
[ RUN      ] EnclaveTest.CpuListComma
tests/enclave_test.cc:274(40625) CHECK FAILED: (*__errno_location ()) != 75 [75 == 75]
errno: 75 [Value too large for defined data type]
PID 40625 Backtrace:
[0] 0x7f5722c6912f : ghost::Exit()
[1] 0x56319013d337 : ghost::(anonymous namespace)::EnclaveTest_CpuListComma_Test::TestBody()
[2] 0x7f57228f6564 : testing::internal::HandleSehExceptionsInMethodIfSupported<>()
[3] 0x7f57228f0af9 : testing::internal::HandleExceptionsInMethodIfSupported<>()
[4] 0x7f57228d7b16 : testing::Test::Run()
[5] 0x7f57228d84ed : testing::TestInfo::Run()
[6] 0x7f57228d8da9 : testing::TestSuite::Run()
[7] 0x7f57228e54d5 : testing::internal::UnitTestImpl::RunAllTests()
[8] 0x7f57228f751e : testing::internal::HandleSehExceptionsInMethodIfSupported<>()
[9] 0x7f57228f1cdb : testing::internal::HandleExceptionsInMethodIfSupported<>()
[10] 0x7f57228e3cc7 : testing::UnitTest::Run()
[11] 0x7f5722943af8 : RUN_ALL_TESTS()
[12] 0x7f5722943a7a : main
[13] 0x7f572224b083 : __libc_start_main
```

**Comment by jackhumphries** [comment]
Are you running these tests directly or with `bazel test`?

**Comment by Swire42** [comment]
`api_test` sometimes pass and sometimes freeze (often at subtest 13, but not always) when ran with `bazel-bin/api_test` or `bazel run api_test`. It seems to always timeouts with `bazel test api_test --spawn_strategy local`.

`bazel-bin/biff_test` shows a libbpf error and then freezes 
during subtests (not always at the same point).

`bazel-bin/database_test` Very slow, 30" per subtest (4' total), succeeds. (timeouts after 60sec with bazel test)

`enclave_test` fails both through bazel and manually with the same error as before.

**Comment by jackhumphries** [comment]
I would just run the tests directly. `bazel test` likely imposes memory constraints on the tests as well.

Let's address the `enclave_test` issue first. Can you add print statements to the test and tell me what the input and output of the `write` syscall is (in `WriteEnclaveCpus()`) when the `EOVERFLOW` error is returned?

**Comment by jackhumphries** [comment]
I wanted to follow up on this. Thanks.

**Comment by Swire42** [comment]
Hi Jack,
Sorry that I missed your messages.
I believe the fastest way to solve these issues would be that you install the public version of ghOSt on a (virtual?) machine of your own.
I am not working with ghOSt actively anymore since my internship involving ghOSt ended in late July, but I still have it installed on my machine, so we can schedule a screen-sharing call if that works better for you.
VM


Note:
From my expericence, the simplest way to install the public version is this :
1) Install Ubuntu 20.04 (any way you want, real or virtual machine)
2) Run the oneliner at https://github.com/Swire42/ghost-installer/blob/master/README.md (installs dependencies (including bazel), compiles the kernel, etc.)
3) Reboot
4) Clone https://github.com/google/ghost-userspace
5) Edit .bazelrc and replace "-std=c++20" by "-std=c++2a" (c++20 not available on Ubuntu 20.04 by default)
6) You should be good now, "bazel build ..." !

**Comment by jackhumphries** [comment]
Thanks. I'm still having trouble reproducing this, so I'll close the issue for now. If someone else has the issue, I can take another look. Thanks for your help.
